File: app/src/main/java/com/bitchat/android/monero/messaging/MoneroMessageHandler.java
package com.bitchat.android.monero.messaging;

import android.util.Log;
import org.json.JSONException;
import org.json.JSONObject;
import java.math.BigInteger;

/**
 * Handles Monero-specific messages and protocols for BitChat
 * Manages payment messages, address sharing, and transaction status updates
 */
public class MoneroMessageHandler {
    
    private static final String TAG = "MoneroMessageHandler";
    
    // Message types
    public static final String TYPE_MONERO_PAYMENT = "monero_payment";
    public static final String TYPE_MONERO_ADDRESS_SHARE = "monero_address_share";
    public static final String TYPE_MONERO_PAYMENT_REQUEST = "monero_payment_request";
    public static final String TYPE_MONERO_PAYMENT_STATUS = "monero_payment_status";
    
    // Payment status values
    public static final String STATUS_PENDING = "pending";
    public static final String STATUS_CONFIRMED = "confirmed";
    public static final String STATUS_FAILED = "failed";
    
    private MoneroMessageListener messageListener;
    
    public interface MoneroMessageListener {
        void onPaymentReceived(MoneroPaymentMessage payment);
        void onAddressShared(String address, String fromUser);
        void onPaymentRequested(MoneroPaymentRequest request);
        void onPaymentStatusUpdated(String txId, String status);
    }
    
    public void setMessageListener(MoneroMessageListener listener) {
        this.messageListener = listener;
    }
    
    /**
     * Parse incoming message and determine if it's Monero-related
     */
    public boolean handleMessage(String messageContent, String fromUser) {
        try {
            JSONObject json = new JSONObject(messageContent);
            String type = json.optString("type", "");
            
            switch (type) {
                case TYPE_MONERO_ADDRESS_SHARE:
                    handleAddressShare(json, fromUser);
                    return true;
                    
                case TYPE_MONERO_PAYMENT_REQUEST:
                    handlePaymentRequest(json, fromUser);
                    return true;
                    
                case TYPE_MONERO_PAYMENT_STATUS:
                    handlePaymentStatus(json, fromUser);
                    return true;
                    
                default:
                    return false; // Not a Monero message
            }
            
        } catch (JSONException e) {
            Log.d(TAG, "Message is not JSON, treating as regular text message");
            return false;
        } catch (Exception e) {
            Log.e(TAG, "Error handling Monero message", e);
            return false;
        }
    }
    
    /**
     * Handle incoming payment message
     */
    private void handlePaymentMessage(JSONObject json, String fromUser) {
        try {
            MoneroPaymentMessage payment = MoneroPaymentMessage.fromJson(json);
            payment.setFromUser(fromUser);
            
            if (messageListener != null) {
                messageListener.onPaymentReceived(payment);
            }
            
            Log.d(TAG, "Received payment: " + payment.getAmount() + " XMR from " + fromUser);
            
        } catch (Exception e) {
            Log.e(TAG, "Error handling payment message", e);
        }
    }
    
    /**
     * Handle address sharing message
     */
    private void handleAddressShare(JSONObject json, String fromUser) {
        try {
            String address = json.getString("address");
            
            if (messageListener != null) {
                messageListener.onAddressShared(address, fromUser);
            }
            
            Log.d(TAG, "Received address share from " + fromUser);
            
        } catch (Exception e) {
            Log.e(TAG, "Error handling address share", e);
        }
    }
    
    /**
     * Handle payment request message
     */
    private void handlePaymentRequest(JSONObject json, String fromUser) {
        try {
            MoneroPaymentRequest request = MoneroPaymentRequest.fromJson(json);
            request.setFromUser(fromUser);
            
            if (messageListener != null) {
                messageListener.onPaymentRequested(request);
            }
            
            Log.d(TAG, "Received payment request from " + fromUser + ": " + request.getAmount() + " XMR");
            
        } catch (Exception e) {
            Log.e(TAG, "Error handling payment request", e);
        }
    }
    
    /**
     * Handle payment status update
     */
    private void handlePaymentStatus(JSONObject json, String fromUser) {
        try {
            String txId = json.getString("tx_id");
            String status = json.getString("status");
            
            if (messageListener != null) {
                messageListener.onPaymentStatusUpdated(txId, status);
            }
            
            Log.d(TAG, "Payment status update: " + txId + " -> " + status);
            
        } catch (Exception e) {
            Log.e(TAG, "Error handling payment status", e);
        }
    }
    
    /**
     * Create a payment message JSON
     */
    public static String createPaymentMessage(String amount, String txId, String recipientAddress) {
        try {
            JSONObject json = new JSONObject();
            json.put("type", TYPE_MONERO_PAYMENT);
            json.put("amount", amount);
            json.put("tx_id", txId);
            json.put("recipient_address", recipientAddress);
            json.put("timestamp", System.currentTimeMillis());
            json.put("status", STATUS_PENDING);
            
            return json.toString();
            
        } catch (JSONException e) {
            Log.e(TAG, "Error creating payment message", e);
            return null;
        }
    }
    
    /**
     * Create an address share message JSON
     */
    public static String createAddressShareMessage(String address) {
        try {
            JSONObject json = new JSONObject();
            json.put("type", TYPE_MONERO_ADDRESS_SHARE);
            json.put("address", address);
            json.put("timestamp", System.currentTimeMillis());
            
            return json.toString();
            
        } catch (JSONException e) {
            Log.e(TAG, "Error creating address share message", e);
            return null;
        }
    }
    
    /**
     * Create a payment request message JSON
     */
    public static String createPaymentRequestMessage(String amount, String reason) {
        try {
            JSONObject json = new JSONObject();
            json.put("type", TYPE_MONERO_PAYMENT_REQUEST);
            json.put("amount", amount);
            json.put("reason", reason != null ? reason : "");
            json.put("timestamp", System.currentTimeMillis());
            
            return json.toString();
            
        } catch (JSONException e) {
            Log.e(TAG, "Error creating payment request message", e);
            return null;
        }
    }
    
    /**
     * Create a payment status update message JSON
     */
    public static String createPaymentStatusMessage(String txId, String status) {
        try {
            JSONObject json = new JSONObject();
            json.put("type", TYPE_MONERO_PAYMENT_STATUS);
            json.put("tx_id", txId);
            json.put("status", status);
            json.put("timestamp", System.currentTimeMillis());
            
            return json.toString();
            
        } catch (JSONException e) {
            Log.e(TAG, "Error creating payment status message", e);
            return null;
        }
    }
    
    /**
     * MoneroPaymentMessage class
     */
    public static class MoneroPaymentMessage {
        private String amount;
        private String txId;
        private String recipientAddress;
        private long timestamp;
        private String status;
        private String fromUser;
        
        public static MoneroPaymentMessage fromJson(JSONObject json) throws JSONException {
            MoneroPaymentMessage message = new MoneroPaymentMessage();
            message.amount = json.getString("amount");
            message.txId = json.getString("tx_id");
            message.recipientAddress = json.getString("recipient_address");
            message.timestamp = json.getLong("timestamp");
            message.status = json.optString("status", STATUS_PENDING);
            return message;
        }
        
        // Getters and setters
        public String getAmount() { return amount; }
        public void setAmount(String amount) { this.amount = amount; }
        
        public String getTxId() { return txId; }
        public void setTxId(String txId) { this.txId = txId; }
        
        public String getRecipientAddress() { return recipientAddress; }
        public void setRecipientAddress(String recipientAddress) { this.recipientAddress = recipientAddress; }
        
        public long getTimestamp() { return timestamp; }
        public void setTimestamp(long timestamp) { this.timestamp = timestamp; }
        
        public String getStatus() { return status; }
        public void setStatus(String status) { this.status = status; }
        
        public String getFromUser() { return fromUser; }
        public void setFromUser(String fromUser) { this.fromUser = fromUser; }
        
        public boolean isPending() { return STATUS_PENDING.equals(status); }
        public boolean isConfirmed() { return STATUS_CONFIRMED.equals(status); }
        public boolean isFailed() { return STATUS_FAILED.equals(status); }
    }
    
    /**
     * MoneroPaymentRequest class
     */
    public static class MoneroPaymentRequest {
        private String amount;
        private String reason;
        private long timestamp;
        private String fromUser;
        
        public static MoneroPaymentRequest fromJson(JSONObject json) throws JSONException {
            MoneroPaymentRequest request = new MoneroPaymentRequest();
            request.amount = json.getString("amount");
            request.reason = json.optString("reason", "");
            request.timestamp = json.getLong("timestamp");
            return request;
        }
        
        // Getters and setters
        public String getAmount() { return amount; }
        public void setAmount(String amount) { this.amount = amount; }
        
        public String getReason() { return reason; }
        public void setReason(String reason) { this.reason = reason; }
        
        public long getTimestamp() { return timestamp; }
        public void setTimestamp(long timestamp) { this.timestamp = timestamp; }
        
        public String getFromUser() { return fromUser; }
        public void setFromUser(String fromUser) { this.fromUser = fromUser; }
    }
    
    /**
     * Utility method to check if a message is Monero-related
     */
    public static boolean isMoneroMessage(String messageContent) {
        try {
            JSONObject json = new JSONObject(messageContent);
            String type = json.optString("type", "");
            return type.startsWith("monero_");
        } catch (JSONException e) {
            return false;
        }
    }
    
    /**
     * Extract display text from Monero message for chat UI
     */
    public static String getDisplayText(String messageContent) {
        try {
            JSONObject json = new JSONObject(messageContent);
            String type = json.optString("type", "");
            
            switch (type) {
                case TYPE_MONERO_PAYMENT:
                    String amount = json.getString("amount");
                    String status = json.optString("status", STATUS_PENDING);
                    return "💰 Sent " + amount + " XMR (" + status + ")";
                    
                case TYPE_MONERO_ADDRESS_SHARE:
                    return "📍 Shared Monero address";
                    
                case TYPE_MONERO_PAYMENT_REQUEST:
                    String reqAmount = json.getString("amount");
                    String reason = json.optString("reason", "");
                    return "💳 Requested " + reqAmount + " XMR" + 
                           (reason.isEmpty() ? "" : " - " + reason);
                    
                case TYPE_MONERO_PAYMENT_STATUS:
                    String txId = json.getString("tx_id");
                    String newStatus = json.getString("status");
                    return "ℹ️ Payment " + txId.substring(0, 8) + "... " + newStatus;
                    
                default:
                    return messageContent;
            }
        } catch (JSONException e) {
            return messageContent;
        }
    }
}

File: app/src/main/java/com/bitchat/android/monero/messaging/MoneroMessageHandler.kt
package com.bitchat.android.monero.messaging

import android.util.Log
import com.bitchat.android.model.BitchatMessage
import org.json.JSONObject

private const val TAG = "MoneroMessageHandler"

/**
 * Unified MoneroMessageHandler
 * Supports:
 * - Legacy blob-based flow ([PREFIX] + raw JSON)
 * - JSON type-based flow {"type":"monero_*", ...}
 */
class MoneroMessageHandler(
    private val onPaymentReceived: (com.bitchat.android.monero.messaging.MoneroMessageHandler.MoneroPaymentMessage) -> Unit,
    private val onAddressShared: (address: String, fromUser: String) -> Unit,
    private val onPaymentRequested: (MoneroPaymentRequest) -> Unit,
    private val onPaymentStatusUpdated: (txId: String, status: String) -> Unit,
    private val onTransactionIdReceived: (TransactionIdMessage) -> Unit,
    private val onTransactionSearchRequested: (SearchTransactionRequest) -> Unit
) {

    interface MoneroMessageListener {
        fun onPaymentReceived(payment: com.bitchat.android.monero.messaging.MoneroMessageHandler.MoneroPaymentMessage)
        fun onAddressShared(address: String, fromUser: String)
        fun onPaymentRequested(request: MoneroPaymentRequest)
        fun onPaymentStatusUpdated(txId: String, status: String)
        fun onTransactionIdReceived(txIdMessage: TransactionIdMessage)
        fun onTransactionSearchRequested(request: SearchTransactionRequest)
    }

    // Prefix-style message types (legacy)
    companion object {
        const val TYPE_PAYMENT = "[MONERO_PAYMENT]"
        const val TYPE_ADDRESS_SHARE = "[MONERO_ADDRESS]"
        const val TYPE_PAYMENT_REQUEST = "[MONERO_REQUEST]"
        const val TYPE_PAYMENT_STATUS = "[MONERO_STATUS]"
        const val TYPE_ADDRESS_REQUEST = "[REQUEST_MONERO_ADDRESS]"

        const val TYPE_TRANSACTION_ID = "[MONERO_TXID]"
        const val TYPE_TRANSACTION_SEARCH = "[MONERO_SEARCH_TX]"
        const val TYPE_TRANSACTION_FOUND = "[MONERO_TX_FOUND]"
        const val TYPE_TRANSACTION_NOT_FOUND = "[MONERO_TX_NOT_FOUND]"

        // JSON type-based message types (new)
        const val TYPE_MONERO_PAYMENT = "monero_payment"
        const val TYPE_MONERO_ADDRESS_SHARE = "monero_address_share"
        const val TYPE_MONERO_PAYMENT_REQUEST = "monero_payment_request"
        const val TYPE_MONERO_PAYMENT_STATUS = "monero_payment_status"

        // Payment statuses
        const val STATUS_PENDING = "pending"
        const val STATUS_CONFIRMED = "confirmed"
        const val STATUS_FAILED = "failed"

        @JvmStatic
        fun isMoneroMessage(content: String): Boolean {
            return try {
                val json = JSONObject(content)
                val type = json.optString("type", "")
                type.startsWith("monero_")
            } catch (e: Exception) {
                false
            }
        }

        @JvmStatic
        fun getDisplayText(content: String): String {
            return try {
                val json = JSONObject(content)
                when (val type = json.optString("type", "")) {
                    TYPE_MONERO_PAYMENT -> {
                        val amount = json.getString("amount")
                        val status = json.optString("status", STATUS_PENDING)
                        "💰 Sent $amount XMR ($status)"
                    }
                    TYPE_MONERO_ADDRESS_SHARE -> "📍 Shared Monero address"
                    TYPE_MONERO_PAYMENT_REQUEST -> {
                        val amount = json.getString("amount")
                        val reason = json.optString("reason", "")
                        "💳 Requested $amount XMR" + if (reason.isNotEmpty()) " - $reason" else ""
                    }
                    TYPE_MONERO_PAYMENT_STATUS -> {
                        val txId = json.getString("tx_id")
                        val status = json.getString("status")
                        "ℹ️ Payment ${txId.take(8)}... $status"
                    }
                    else -> content
                }
            } catch (e: Exception) {
                content
            }
        }
        
        @JvmStatic
        fun initializeMoneroMessageHandler(
            listener: MoneroMessageListener
        ): MoneroMessageHandler {
            return MoneroMessageHandler(
                onPaymentReceived = { payment -> 
                    listener.onPaymentReceived(payment)
                    Unit // Add explicit Unit return to fix "missing return value" error
                },
                onAddressShared = { addr, from -> 
                    listener.onAddressShared(addr, from)
                    Unit
                },
                onPaymentRequested = { request -> 
                    listener.onPaymentRequested(request)
                    Unit
                },
                onPaymentStatusUpdated = { txId, status -> 
                    listener.onPaymentStatusUpdated(txId, status)
                    Unit
                },
                onTransactionIdReceived = { txIdMsg -> 
                    listener.onTransactionIdReceived(txIdMsg)
                    Unit
                },
                onTransactionSearchRequested = { request -> 
                    listener.onTransactionSearchRequested(request)
                    Unit
                }
            ).also {
                Log.i("MoneroWalletHandler", "✅ MoneroMessageHandler initialized")
            }
        }      
    }

    // Data classes
    data class MoneroPaymentMessage(
        val amount: String,
        val fromUser: String,
        val signedTxBlob: String? = null,
        val txId: String? = null,
        val toAddress: String? = null,
        val timestamp: Long,
        val status: String = STATUS_PENDING
    )

    data class MoneroPaymentRequest(
        val amount: String,
        val fromUser: String,
        val reason: String,
        val timestamp: Long
    )

    data class TransactionIdMessage(
        val txId: String,
        val amount: String,
        val fromUser: String,
        val toAddress: String,
        val timestamp: Long,
        val blockHeight: Long = 0
    ) {
        companion object {
            fun createDefault(
                txId: String,
                fromUser: String,
                amount: String = "0",
                toAddress: String = "",
                blockHeight: Long = 0
            ): TransactionIdMessage {
                return TransactionIdMessage(
                    txId = txId,
                    amount = amount,
                    fromUser = fromUser,
                    toAddress = toAddress,
                    timestamp = System.currentTimeMillis(),
                    blockHeight = blockHeight
                )
            }
        }
    }


    data class SearchTransactionRequest(
        val txId: String,
        val query: String,
        val fromUser: String,
        val timestamp: Long
    )

    data class TransactionFoundMessage(
        val txId: String,
        val amount: String,
        val confirmations: Long,
        val blockHeight: Long
    )
    
    /**
     * Main message processor
     */
    fun processMoneroMessage(message: BitchatMessage): Boolean {
        val content = message.content

        // Legacy [PREFIX]-style parsing
        when {
            content.startsWith(TYPE_PAYMENT) -> {
                handlePaymentMessage(content, message.sender)
                return true
            }
            content.startsWith(TYPE_ADDRESS_SHARE) -> {
                handleAddressShare(content, message.sender)
                return true
            }
            content.startsWith(TYPE_PAYMENT_REQUEST) -> {
                handlePaymentRequest(content, message.sender)
                return true
            }
            content.startsWith(TYPE_PAYMENT_STATUS) -> {
                handlePaymentStatus(content)
                return true
            }
            content.startsWith(TYPE_TRANSACTION_ID) -> {
                handleTransactionId(content, message.sender)
                return true
            }
            content.startsWith(TYPE_TRANSACTION_SEARCH) -> {
                handleTransactionSearch(content, message.sender)
                return true
            }
            content.startsWith(TYPE_TRANSACTION_FOUND) -> {
                handleTransactionFound(content)
                return true
            }
            content.startsWith(TYPE_TRANSACTION_NOT_FOUND) -> {
                handleTransactionNotFound(content)
                return true
            }
        }

        // JSON type-based parsing
        return try {
            val json = JSONObject(content)
            when (val type = json.optString("type", "")) {
                TYPE_MONERO_PAYMENT -> {
                    val payment = MoneroPaymentMessage(
                        amount = json.getString("amount"),
                        txId = json.optString("tx_id", null),
                        toAddress = json.optString("recipient_address", null),
                        fromUser = message.sender,
                        signedTxBlob = json.optString("signed_tx", null),
                        timestamp = json.getLong("timestamp"),
                        status = json.optString("status", STATUS_PENDING)
                    )
                    onPaymentReceived(payment)
                    true
                }
                TYPE_MONERO_ADDRESS_SHARE -> {
                    val address = json.getString("address")
                    onAddressShared(address, message.sender)
                    true
                }
                TYPE_MONERO_PAYMENT_REQUEST -> {
                    val request = MoneroPaymentRequest(
                        amount = json.getString("amount"),
                        reason = json.optString("reason", ""),
                        timestamp = json.getLong("timestamp"),
                        fromUser = message.sender
                    )
                    onPaymentRequested(request)
                    true
                }
                TYPE_MONERO_PAYMENT_STATUS -> {
                    val txId = json.getString("tx_id")
                    val status = json.getString("status")
                    onPaymentStatusUpdated(txId, status)
                    true
                }
                else -> false
            }
        } catch (e: Exception) {
            false
        }
    }

    // Legacy handlers
    private fun handlePaymentMessage(content: String, sender: String) {
        try {
            val json = JSONObject(content.substring(TYPE_PAYMENT.length))
            val payment = MoneroPaymentMessage(
                amount = json.getString("amount"),
                fromUser = sender,
                signedTxBlob = json.getString("signed_tx"),
                timestamp = json.getLong("timestamp")
            )
            onPaymentReceived(payment)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse payment message: ${e.message}")
        }
    }

    private fun handleAddressShare(content: String, sender: String) {
        try {
            val json = JSONObject(content.substring(TYPE_ADDRESS_SHARE.length))
            val address = json.getString("address")
            onAddressShared(address, sender)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse address share: ${e.message}")
        }
    }

    private fun handlePaymentRequest(content: String, sender: String) {
        try {
            val json = JSONObject(content.substring(TYPE_PAYMENT_REQUEST.length))
            val request = MoneroPaymentRequest(
                amount = json.getString("amount"),
                fromUser = sender,
                reason = json.optString("reason", ""),
                timestamp = json.getLong("timestamp")
            )
            onPaymentRequested(request)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse payment request: ${e.message}")
        }
    }

    private fun handlePaymentStatus(content: String) {
        try {
            val json = JSONObject(content.substring(TYPE_PAYMENT_STATUS.length))
            val txId = json.getString("tx_id")
            val status = json.getString("status")
            onPaymentStatusUpdated(txId, status)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse payment status: ${e.message}")
        }
    }

    private fun handleTransactionId(content: String, sender: String) {
        try {
            val json = JSONObject(content.substring(TYPE_TRANSACTION_ID.length))
            val txIdMessage = TransactionIdMessage(
                txId = json.getString("tx_id"),
                amount = json.getString("amount"),
                fromUser = sender,
                toAddress = json.getString("to_address"),
                timestamp = json.getLong("timestamp"),
                blockHeight = json.optLong("block_height", 0)
            )
            onTransactionIdReceived(txIdMessage)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse transaction ID message: ${e.message}")
        }
    }

    private fun handleTransactionSearch(content: String, sender: String) {
        try {
            val json = JSONObject(content.substring(TYPE_TRANSACTION_SEARCH.length))
            val searchRequest = SearchTransactionRequest(
                txId = json.getString("tx_id"),
                query = content,
                fromUser = sender,
                timestamp = json.getLong("timestamp")
            )
            onTransactionSearchRequested(searchRequest)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse transaction search request: ${e.message}")
        }
    }

    private fun handleTransactionFound(content: String) {
        try {
            val json = JSONObject(content.substring(TYPE_TRANSACTION_FOUND.length))
            val txId = json.getString("tx_id")
            val status = "Found: ${json.getInt("confirmations")} confirmations"
            onPaymentStatusUpdated(txId, status)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse transaction found message: ${e.message}")
        }
    }

    private fun handleTransactionNotFound(content: String) {
        try {
            val json = JSONObject(content.substring(TYPE_TRANSACTION_NOT_FOUND.length))
            val txId = json.getString("tx_id")
            onPaymentStatusUpdated(txId, "Not found")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse transaction not found message: ${e.message}")
        }
    }

    // Builders (both legacy + JSON)
    fun createPaymentMessage(amount: String, signedTxBlob: String): String {
        val json = JSONObject().apply {
            put("amount", amount)
            put("signed_tx", signedTxBlob)
            put("timestamp", System.currentTimeMillis())
        }
        return "$TYPE_PAYMENT$json"
    }

    fun createAddressShareMessage(address: String): String {
        val json = JSONObject().apply {
            put("address", address)
            put("timestamp", System.currentTimeMillis())
        }
        return "$TYPE_ADDRESS_SHARE$json"
    }

    fun createPaymentRequestMessage(amount: String, reason: String = ""): String {
        val json = JSONObject().apply {
            put("amount", amount)
            put("reason", reason)
            put("timestamp", System.currentTimeMillis())
        }
        return "$TYPE_PAYMENT_REQUEST$json"
    }

    fun createPaymentStatusMessage(txId: String, status: String): String {
        val json = JSONObject().apply {
            put("tx_id", txId)
            put("status", status)
            put("timestamp", System.currentTimeMillis())
        }
        return "$TYPE_PAYMENT_STATUS$json"
    }

    fun createTransactionIdMessage(
        txId: String,
        amount: String,
        toAddress: String,
        blockHeight: Long = 0
    ): String {
        val json = JSONObject().apply {
            put("tx_id", txId)
            put("amount", amount)
            put("to_address", toAddress)
            put("timestamp", System.currentTimeMillis())
            put("block_height", blockHeight)
        }
        return "$TYPE_TRANSACTION_ID$json"
    }

    fun createTransactionSearchMessage(txId: String): String {
        val json = JSONObject().apply {
            put("tx_id", txId)
            put("timestamp", System.currentTimeMillis())
        }
        return "$TYPE_TRANSACTION_SEARCH$json"
    }

    fun createTransactionFoundMessage(
        txId: String,
        amount: String,
        confirmations: Long,
        blockHeight: Long
    ): String {
        val json = JSONObject().apply {
            put("tx_id", txId)
            put("amount", amount)
            put("confirmations", confirmations)
            put("block_height", blockHeight)
            put("timestamp", System.currentTimeMillis())
        }
        return "$TYPE_TRANSACTION_FOUND$json"
    }

    fun createTransactionNotFoundMessage(txId: String): String {
        val json = JSONObject().apply {
            put("tx_id", txId)
            put("timestamp", System.currentTimeMillis())
        }
        return "$TYPE_TRANSACTION_NOT_FOUND$json"
    }
}

